pkg/skaffold/build/jib/sync.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jib
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+	"regexp"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+type SyncMap map[string]SyncEntry
+
+type SyncEntry struct {
+	Dest     []string
+	FileTime time.Time
+	IsDirect bool
+}
+
+type JSONSyncMap struct {
+	Direct    []JSONSyncEntry `json:"direct"`
+	Generated []JSONSyncEntry `json:"generated"`
+}
+
+type JSONSyncEntry struct {
+	Src  string `json:"src"`
+	Dest string `json:"dest"`
+}
+
+func getSyncMapFromSystem(cmd *exec.Cmd) (*SyncMap, error) {
+	jsm := JSONSyncMap{}
+	stdout, err := util.RunCmdOut(cmd)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get Jib sync map")
+	}
+
+	// To parse the output, search for "BEGIN JIB JSON", then unmarshal the next line into the pathMap struct.
+	// Syncmap is transitioning to "BEGIN JIB JSON: SYNCMAP/1" starting in jib 2.0.0
+	// perhaps this feature should only be included from 2.0.0 onwards? And we generally avoid this?
+	matches := regexp.MustCompile(`BEGIN JIB JSON(?:: SYNCMAP/1)?\r?\n({.*})`).FindSubmatch(stdout)
```

Comment:
Do you think it's better to fix the version `/1`? For example, if the user users a newer Jib version returning `/2`, the user has to upgrade Skaffold too. Of course, that can be our strategy, but the error message below when there's no match doesn't help to the user. The user will just see "failed to get Jib Sync data" after upgrading Jib.

And perhaps remove the following comment? Weird to leave it in the code.

``// perhaps this feature should only be included from 2.0.0 onwards? And we generally avoid this?``

Reply:
I want to leave it in while we are still doing both types of parsing. (comment will be removed and code cleaned up after we 2.0.0 and we have a way to variably set the value of: https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/jib/maven.go#L39 -- which will happen in a followup)

So doing `SYNCMAP/version` is just providing us some metadata for futureproofing. We don't actually use this number in anyway, but for instance if we do create a new version of skaffold that accepts both 1 and 2, we can make this value different. Currently, if the API changes, the parser will die anyway.

Reply:
I guess we are not on the same page...

For example, in 2020, imagine we release Skaffold version A having this code that looks for `SYNCMAP/1`. In 2021, we release Skaffold version B that can parse both `/1` and `/2` and a new Jib version that returns `SYNCMAP/2`. Now, at some point, the user upgrades Jib to a new version (returning `/2`), but the user is still using Skaffold version A. Then, my concern is that `skaffold dev` will just die saying

```
error: failed to get Jib Sync data
```

And the user has no idea why.

Reply:
But I guess we'll have some sort of problem either way, so I'm not actually against settling with this policy.

Reply:
Right, I think at least this way it possible to make it a little more explicit.